core/src/query.rs

```diff
@@ -1,77 +1,551 @@
+                }}];
+
+                let options = query_timeout.map(|i| {
+                    AggregateOptions::builder()
+                        .max_time(Duration::from_millis(i as u64))
+                        .build()
+                });
+
+                let cursor = db.aggregate(pipeline, options)?;
+
+                Ok(MongoQuery {
+                    resultset_cursor: cursor,
+                    resultset_metadata: metadata,
+                    current: None,
+                })
+            }
+        }
     }
 
     // Return the number of fields/columns in the resultset
     fn _get_col_count(&self) -> u32 {
-        unimplemented!()
+        self.resultset_metadata.len() as u32
     }
 
     // Get the metadata for the column with the given index.
-    fn _get_col_metadata(&self, _col_index: u16) -> Result<MongoColMetadata> {
-        unimplemented!()
+    fn get_col_metadata(&self, col_index: u16) -> Result<&MongoColMetadata> {
+        self.resultset_metadata
+            .get(col_index as usize)
+            .map_or(Err(Error::ColIndexOutOfBounds(col_index)), |md| Ok(md))
     }
 }
 
 impl MongoStatement for MongoQuery {
     // Move the cursor to the next document and update the current row.
     // Return true if moving was successful, false otherwise.
+    // This method deserializes the current row and stores it in self.
     fn next(&mut self) -> Result<bool> {
-        unimplemented!()
+        let res = self.resultset_cursor.advance().map_err(Error::Mongo);
+        self.current = Some(self.resultset_cursor.deserialize_current()?);
+        res
     }
 
     // Get the BSON value for the cell at the given colIndex on the current row.
     // Fails if the first row as not been retrieved (next must be called at least once before getValue).
-    fn get_value(&self, _col_index: u16) -> Result<Option<Bson>> {
-        unimplemented!()
+    fn get_value(&self, col_index: u16) -> Result<Option<Bson>> {
+        let current = self.current.as_ref().ok_or(Error::InvalidCursorState)?;
+        let md = self.get_col_metadata(col_index)?;
+        let datasource = current
+            .get_document(&md.table_name)
+            .map_err(Error::ValueAccess)?;
+        let column = datasource.get(&md.col_name);
+        Ok(column.cloned())
     }
 }
 
 // Metadata information for a column of the result set.
 // The information is to be used when reporting columns information from
-// SQLColAtrribute or SQLDescibeCol and when converting the data to the targeted C type.
-#[derive(Debug)]
+// SQLColAttribute or SQLDescribeCol and when converting the data to the targeted C type.
+#[derive(Clone, Debug)]
 pub struct MongoColMetadata {
     pub base_col_name: String,
     pub base_table_name: String,
     pub catalog_name: String,
-    pub col_count: u16,
-    pub display_size: u64,
+    pub display_size: Option<u16>,
     pub fixed_prec_scale: bool,
     pub label: String,
-    pub length: u128,
+    pub length: Option<u16>,
```

Comment:
Several of these were set to `null` in the JDBC driver and can be "undefined" according to the ODBC spec. I figured making such fields optional was the best way to capture that.